File: js/fndsa/test/fndsa.test.js
import { FNDSA512, FNDSA1024, FNDSAPadded512, FNDSAPadded1024, keyGen, sign, verify } from '../src/index.js';
import { readFileSync, existsSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

describe('FN-DSA', () => {
  test.each([FNDSA512, FNDSA1024, FNDSAPadded512, FNDSAPadded1024])('param sizes %s', (p) => {
    expect(p.pkSize).toBeGreaterThan(0);
    expect(p.skSize).toBeGreaterThan(0);
    expect(p.sigSize).toBeGreaterThan(0);
    expect(p.betaSq).toBeGreaterThan(0);
  });

  test('param values FNDSA512', () => {
    expect(FNDSA512.n).toBe(512);
    expect(FNDSA512.logN).toBe(9);
    expect(FNDSA512.pkSize).toBe(897);
    expect(FNDSA512.skSize).toBe(1281);
    expect(FNDSA512.sigSize).toBe(666);
    expect(FNDSA512.padded).toBe(false);
  });

  test('param values FNDSA1024', () => {
    expect(FNDSA1024.n).toBe(1024);
    expect(FNDSA1024.logN).toBe(10);
    expect(FNDSA1024.pkSize).toBe(1793);
    expect(FNDSA1024.skSize).toBe(2305);
    expect(FNDSA1024.sigSize).toBe(1280);
    expect(FNDSA1024.padded).toBe(false);
  });

  // Only run 512 in tests (1024 is slow)
  test('roundtrip FNDSA512', () => {
    const [pk, sk] = keyGen(FNDSA512);
    expect(pk.length).toBe(FNDSA512.pkSize);
    expect(sk.length).toBe(FNDSA512.skSize);
    const msg = Buffer.from('test message fn-dsa');
    const sig = sign(sk, msg, FNDSA512);
    expect(sig.length).toBeLessThanOrEqual(FNDSA512.sigSize);
    expect(verify(pk, msg, sig, FNDSA512)).toBe(true);
    expect(verify(pk, Buffer.from('wrong'), sig, FNDSA512)).toBe(false);
    const tampered = Buffer.from(sig);
    tampered[Math.min(42, tampered.length - 1)] ^= 0x01;
    expect(verify(pk, msg, tampered, FNDSA512)).toBe(false);
  }, 60000);

  test('roundtrip FNDSAPadded512', () => {
    const [pk, sk] = keyGen(FNDSAPadded512);
    expect(pk.length).toBe(FNDSAPadded512.pkSize);
    expect(sk.length).toBe(FNDSAPadded512.skSize);
    const msg = Buffer.from('test message fn-dsa padded');
    const sig = sign(sk, msg, FNDSAPadded512);
    expect(sig.length).toBe(FNDSAPadded512.sigSize);
    expect(verify(pk, msg, sig, FNDSAPadded512)).toBe(true);
  }, 60000);

  test('interop vectors', () => {
    let anyRan = false;
    for (const [name, params] of [['FN-DSA-512', FNDSA512], ['FN-DSA-1024', FNDSA1024]]) {
      const path = join(__dirname, `../../../test-vectors/fn-dsa/${name}.json`);
      if (!existsSync(path)) continue;
      const { vectors } = JSON.parse(readFileSync(path, 'utf8'));
      for (const v of vectors) {
        const pk = Buffer.from(v.pk, 'hex');
        const msg = Buffer.from(v.msg, 'hex');
        const sig = Buffer.from(v.sig, 'hex');
        expect(verify(pk, msg, sig, params)).toBe(true);
      }
      anyRan = true;
    }
    if (!anyRan) return; // no test vector files found, skip
  });
});

File: java/src/test/java/com/pqc/fndsa/FnDsaTest.java
package com.pqc.fndsa;

import org.junit.jupiter.api.*;
import org.junit.jupiter.params.*;
import org.junit.jupiter.params.provider.*;

import java.nio.file.Files;
import java.nio.file.Path;
import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.HexFormat;
import java.util.List;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

class FnDsaTest {

    @ParameterizedTest
    @MethodSource("allParams")
    void testParamSizes(Params p) {
        assertTrue(p.pkSize > 0, "pkSize should be positive for " + p.name);
        assertTrue(p.skSize > 0, "skSize should be positive for " + p.name);
        assertTrue(p.sigSize > 0, "sigSize should be positive for " + p.name);
        assertTrue(p.betaSq > 0, "betaSq should be positive for " + p.name);
    }

    @ParameterizedTest
    @MethodSource("allParams")
    void testRoundtrip(Params p) {
        SecureRandom rng = new SecureRandom();
        byte[][] keys = FnDsa.keyGen(p, rng);
        byte[] pk = keys[0];
        byte[] sk = keys[1];

        assertEquals(p.pkSize, pk.length, "pk size mismatch for " + p.name);
        assertEquals(p.skSize, sk.length, "sk size mismatch for " + p.name);

        byte[] msg = "test message FN-DSA".getBytes();
        byte[] sig = FnDsa.sign(sk, msg, p, rng);

        assertNotNull(sig, "signature should not be null for " + p.name);
        if (p.padded) {
            assertEquals(p.sigSize, sig.length, "padded sig size mismatch for " + p.name);
        } else {
            assertTrue(sig.length <= p.sigSize, "sig exceeds max size for " + p.name);
        }

        assertTrue(FnDsa.verify(pk, msg, sig, p), "valid signature should verify for " + p.name);
        assertFalse(FnDsa.verify(pk, "wrong".getBytes(), sig, p),
            "wrong message should not verify for " + p.name);

        byte[] tampered = sig.clone();
        tampered[Math.min(42, tampered.length - 1)] ^= 0x01;
        assertFalse(FnDsa.verify(pk, msg, tampered, p),
            "tampered signature should not verify for " + p.name);
    }

    @Test
    void testInteropVectors() throws Exception {
        HexFormat hex = HexFormat.of();
        for (Object[] entry : new Object[][] {
            {"FN-DSA-512", Params.FNDSA512},
            {"FN-DSA-1024", Params.FNDSA1024}
        }) {
            String name = (String) entry[0];
            Params p = (Params) entry[1];

            // Try multiple paths (from different working directories)
            Path path = null;
            for (String candidate : new String[] {
                "../test-vectors/fn-dsa/" + name + ".json",
                "../../test-vectors/fn-dsa/" + name + ".json",
                "../../../test-vectors/fn-dsa/" + name + ".json",
                "test-vectors/fn-dsa/" + name + ".json"
            }) {
                Path cp = Path.of(candidate);
                if (Files.exists(cp)) {
                    path = cp;
                    break;
                }
            }

            if (path == null) {
                System.out.println("Skipping " + name + " -- vector file not found");
                continue;
            }

            String json = Files.readString(path);
            List<TestVector> vectors = parseVectors(json);
            System.out.println("Testing " + vectors.size() + " vectors for " + name);

            for (int i = 0; i < vectors.size(); i++) {
                TestVector v = vectors.get(i);
                byte[] pk = hex.parseHex(v.pk);
                byte[] msg = hex.parseHex(v.msg);
                byte[] sig = hex.parseHex(v.sig);

                assertTrue(FnDsa.verify(pk, msg, sig, p),
                    "Interop vector " + i + " for " + name + " should verify");
            }
        }
    }

    // Minimal JSON parser for test vectors (no Jackson dependency needed)
    private static List<TestVector> parseVectors(String json) {
        List<TestVector> result = new ArrayList<>();
        // Find "vectors" array
        int vectorsIdx = json.indexOf("\"vectors\"");
        if (vectorsIdx < 0) return result;
        int arrStart = json.indexOf('[', vectorsIdx);
        if (arrStart < 0) return result;

        int pos = arrStart + 1;
        while (pos < json.length()) {
            // Skip whitespace
            while (pos < json.length() && Character.isWhitespace(json.charAt(pos))) pos++;
            if (pos >= json.length() || json.charAt(pos) == ']') break;
            if (json.charAt(pos) != '{') { pos++; continue; }

            // Find matching }
            int depth = 0;
            int objStart = pos;
            int objEnd = pos;
            for (int i = pos; i < json.length(); i++) {
                char ch = json.charAt(i);
                if (ch == '{') depth++;
                else if (ch == '}') {
                    depth--;
                    if (depth == 0) { objEnd = i; break; }
                }
            }
            String obj = json.substring(objStart, objEnd + 1);
            TestVector tv = new TestVector();
            tv.pk = extractField(obj, "pk");
            tv.msg = extractField(obj, "msg");
            tv.sig = extractField(obj, "sig");
            if (tv.pk != null && tv.msg != null && tv.sig != null) {
                result.add(tv);
            }
            pos = objEnd + 1;
        }
        return result;
    }

    private static String extractField(String obj, String field) {
        String key = "\"" + field + "\"";
        int idx = obj.indexOf(key);
        if (idx < 0) return null;
        int colonIdx = obj.indexOf(':', idx + key.length());
        if (colonIdx < 0) return null;
        int quoteStart = obj.indexOf('"', colonIdx + 1);
        if (quoteStart < 0) return null;
        int quoteEnd = obj.indexOf('"', quoteStart + 1);
        if (quoteEnd < 0) return null;
        return obj.substring(quoteStart + 1, quoteEnd);
    }

    private static class TestVector {
        String pk, msg, sig;
    }

    static Stream<Params> allParams() {
        return Stream.of(
            Params.FNDSA512,
            Params.FNDSA1024,
            Params.FNDSA_PADDED_512,
            Params.FNDSA_PADDED_1024
        );
    }
}
